conda/models/match_spec.py

```diff
@@ -404,6 +406,36 @@ def fn(self):
         assert val
         return val
 
+    @classmethod
+    def merge(cls, match_specs):
+        match_specs = tuple(cls(s) for s in match_specs)
+        grouped = groupby(lambda spec: spec.get_exact_value('name'), match_specs)
+        specs_map = {
+            name: reduce(lambda x, y: x._merge(y), specs) if len(specs) > 1 else specs[0]
+            for name, specs in iteritems(grouped)
+        }
+        return tuple(itervalues(specs_map))
+
+    def _merge(self, other):
+        if self.optional != other.optional or self.target != other.target:
+            raise ValueError("Incompatible MatchSpec merge:  - %s\n  - %s" % (self, other))
+
+        final_components = {}
+        component_names = set(self._match_components) | set(other._match_components)
+        for component_name in component_names:
+            this_component = self._match_components.get(component_name)
+            that_component = other._match_components.get(component_name)
+            if this_component is None and that_component is None:
+                continue
+            elif this_component is None:
+                final_components[component_name] = that_component
+            elif that_component is None:
+                final_components[component_name] = this_component
+            else:
+                final_components[component_name] = this_component.merge(that_component)
```

Comment:
You could simply this as
```python
final_components = {}
for name, component in iteritems(self.match_components):
    if component is None:
        continue
    final_components[name] = component
for name, component in iteritems(other.match_components):
    if component is None:
        continue
    if name in final_components:
        final_components[name].merge(component)
    else:
        final_components[name] = component
```
or
```python
final_components = {name: comp for name, comp in iteritems(self._match_components) if comp is not None}
for name, component in ((name: comp) for name, comp in iteritems(other._match_components) if comp is not None):
    if name in final_components:
        final_components[name].merge(component)
    else:
        final_components[name] = component
```

Reply:
I plugged both of these in.  They both have four test failures with the additional tests I added after the most recent commit.

Reply:
Writing code directly in GitHub is a **bad** habit...
Sorry, forgot the assignment in
```python
final_components[name] = final_components[name].merge(component)
```
Edited the comment above with that fix.